cranelift/frontend/src/frontend.rs

```diff
@@ -163,11 +164,85 @@ impl<'short, 'long> InstBuilderBase<'short> for FuncInstBuilder<'short, 'long> {
 }
 
 #[derive(Debug, Copy, Clone, PartialEq, Eq)]
-/// This is the error variant returned from [`FunctionBuilder::try_use_var`].
+/// An error encountered when calling [`FunctionBuilder::try_use_var`].
 pub enum UseVariableError {
     UsedBeforeDeclared(Variable),
 }
 
+impl fmt::Display for UseVariableError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            UseVariableError::UsedBeforeDeclared(variable) => {
+                f.write_str("variable ")?;
+                <usize as fmt::Display>::fmt(&variable.index(), f)?;
+                f.write_str(" was used before it was defined!")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl std::error::Error for UseVariableError {}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+/// An error encountered when calling [`FunctionBuilder::try_declare_var`]
+pub enum DeclareVariableError {
+    DeclaredMultipleTimes(Variable),
+}
+
+impl std::error::Error for DeclareVariableError {}
+
+impl fmt::Display for DeclareVariableError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            DeclareVariableError::DeclaredMultipleTimes(variable) => {
+                f.write_str("variable ")?;
+                <usize as fmt::Display>::fmt(&variable.index(), f)?;
+                f.write_str(" was declared multiple times!")?;
```

Comment:
Here you should be able to do `write!(f, "variable {} was declared multiple times", variable.index())`.